# requirements.R
# Install required R packages for OSAHS-Microbiome-Metabolome-Analysis pipeline

# Function to install packages from CRAN
install_cran <- function(packages) {
  new_packages <- packages[!(packages %in% installed.packages()[,"Package"])]
  if(length(new_packages)) install.packages(new_packages)
}

# Function to install packages from Bioconductor
install_bioconductor <- function(packages) {
  if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
  
  new_packages <- packages[!(packages %in% installed.packages()[,"Package"])]
  if(length(new_packages)) BiocManager::install(new_packages)
}

# Function to install packages from GitHub
install_github <- function(packages) {
  if (!requireNamespace("devtools", quietly = TRUE))
    install.packages("devtools")
  
  for (pkg in packages) {
    if (!pkg %in% installed.packages()[,"Package"]) {
      devtools::install_github(pkg)
    }
  }
}

# Install CRAN packages
cran_packages <- c(
  "tidyverse",
  "ggplot2",
  "dplyr",
  "tibble",
  "reshape2",
  "gridExtra",
  "grid",
  "pheatmap",
  "corrplot",
  "Hmisc",
  "RColorBrewer",
  "viridis",
  "formattable",
  "kableExtra",
  "knitr",
  "markdown",
  "data.table",
  "stringr",
  "magrittr",
  "scales",
  "cowplot",
  "ggrepel",
  "networkD3",
  "htmlwidgets",
  "flashClust",
  "WGCNA",
  "factoextra",
  "FactoMineR",
  "vegan",
  "ade4",
  "gridGraphics",
  "htmlTable",
  "tableone",
  "car",
  "lmtest",
  "sandwich",
  "multcomp",
  "emmeans",
  "lme4",
  "nlme",
  "glmmTMB",
  "pscl",
  "MASS",
  "fitdistrplus",
  "flexmix",
  "nnet",
  "partykit",
  "rpart",
  "randomForest",
  "gbm",
  "xgboost",
  "caret",
  "pROC",
  "ROCR",
  "e1071",
  "klaR",
  "kernlab",
  "nnet",
  "pls",
  "glmnet",
  "Matrix",
  "irlba",
  "RSpectra",
  "scatterplot3d",
  "rgl",
  "viridisLite",
  "Rtsne",
  "umap",
  "clustree",
  "dbscan",
  "fpc",
  "mclust",
  "cluster",
  "NbClust",
  "clValid",
  "factoextra",
  "gridBase",
  "hexbin",
  "gplots",
  "RColorBrewer",
  "colorspace",
  "dendextend",
  "circlize",
  "ComplexHeatmap",
  "igraph",
  "ggraph",
  "tidygraph",
  "network",
  "sna",
  "intergraph",
  "visNetwork",
  "threejs",
  "RNeo4j",
  "neo4r",
  "rjson",
  "jsonlite",
  "httr",
  "XML",
  "XML2R",
  "RCurl",
  "curl",
  "data.tree",
  "DiagrammeR",
  "DiagrammeRsvg",
  "rsvg",
  "webshot",
  "htmltools",
  "DT",
  "shiny",
  "shinydashboard",
  "shinythemes",
  "shinyWidgets",
  "shinyjs",
  "shinyBS",
  "shinyalert",
  "colourpicker",
  "plotly",
  "highcharter",
  "dygraphs",
  "metricsgraphics",
  "rbokeh",
  "ggiraph",
  "ggiraphExtra",
  "ggiraphExtra",
  "officer",
  "flextable",
  "rvg",
  "magick",
  "webshot",
  "png",
  "jpeg",
  "tiff",
  "pdftools",
  "qpdf",
  "zip",
  "tar",
  "openssl",
  "sodium",
  "keyring",
  "credentials",
  "git2r",
  "githubinstall",
  "devtools",
  "roxygen2",
  "testthat",
  "knitr",
  "rmarkdown",
  "bookdown",
  "pkgdown",
  "blogdown",
  "flexdashboard",
  "learnr",
  "tufte",
  "xaringan",
  "rticles",
  "tinytex",
  "latex2exp",
  "tikzDevice",
  "R2HTML",
  "markdown",
  "pander",
  "xtable",
  "stargazer",
  "texreg",
  "apsrtable",
  "memisc",
  "tables",
  "reporttools",
  "knitcitations",
  "bibtex",
  "RefManageR",
  "citr",
  "rbibutils",
  "roxygen2md",
  "docopt",
  "argparse",
  "optparse",
  "config",
  "yaml",
  "jsonlite",
  "xml2",
  "xml",
  "rvest",
  "httr2",
  "curl",
  "RCurl",
  "httpuv",
  "websockets",
  "later",
  "promises",
  "fastmap",
  "htmlwidgets",
  "htmltools",
  "shiny",
  "shinydashboard",
  "shinythemes",
  "shinyWidgets",
  "shinyjs",
  "shinyBS",
  "shinyalert",
  "colourpicker",
  "plotly",
  "highcharter",
  "dygraphs",
  "metricsgraphics",
  "rbokeh",
  "ggiraph",
  "ggiraphExtra",
  "officer",
  "flextable",
  "rvg",
  "magick",
  "webshot",
  "png",
  "jpeg",
  "tiff",
  "pdftools",
  "qpdf",
  "zip",
  "tar",
  "openssl",
  "sodium",
  "keyring",
  "credentials",
  "git2r",
  "githubinstall",
  "devtools",
  "roxygen2",
  "testthat"
)

install_cran(cran_packages)

# Install Bioconductor packages
bioconductor_packages <- c(
  "phyloseq",
  "microbiome",
  "DESeq2",
  "edgeR",
  "limma",
  "impute",
  "preprocessCore",
  "sva",
  "ComBat",
  "pcaMethods",
  "RUVSeq",
  "Biobase",
  "SummarizedExperiment",
  "GenomicRanges",
  "IRanges",
  "S4Vectors",
  "Biostrings",
  "ShortRead",
  "GenomicAlignments",
  "Rsamtools",
  "GenomicFeatures",
  "TxDb.Hsapiens.UCSC.hg38.knownGene",
  "org.Hs.eg.db",
  "clusterProfiler",
  "DOSE",
  "pathview",
  "ReactomePA",
  "gage",
  "gageData",
  "fgsea",
  "GSVA",
  "plsgenomics",
  "mixOmics",
  "ANCOMBC",
  "corncob",
  "Maaslin2",
  "metagenomeSeq",
  "lefser",
  "metagenomeseq",
  "ALDex2",
  "metap",
  "qvalue",
  "fdrtool",
  "qvalue",
  "ashr",
  "locfdr",
  "marray",
  "affy",
  "oligo",
  "arrayQualityMetrics",
  "vsn",
  "normexpr",
  "RColorBrewer",
  "BiocStyle",
  "knitcitations",
  "BiocManager",
  "BiocGenerics",
  "BiocVersion",
  "BiocInstaller",
  "BiocParallel",
  "BiocIO",
  "BiocFileCache",
  "AnnotationTools",
  "annotate",
  "GO.db",
  "KEGGREST",
  "biomaRt",
  "ensembldb",
  "AnnotationDbi",
  "AnnotationHub",
  "ExperimentHub",
  "TxDb.Mmusculus.UCSC.mm10.knownGene",
  "org.Mm.eg.db",
  "TxDb.Rnorvegicus.UCSC.rn6.knownGene",
  "org.Rn.eg.db",
  "TxDb.Dmelanogaster.UCSC.dm6.ensGene",
  "org.Dm.eg.db",
  "TxDb.Celegans.UCSC.ce11.ensGene",
  "org.Ce.eg.db",
  "TxDb.Scerevisiae.UCSC.sacCer3.sgdGene",
  "org.Sc.sgd.db",
  "BSgenome",
  "BSgenome.Hsapiens.UCSC.hg38",
  "BSgenome.Mmusculus.UCSC.mm10",
  "BSgenome.Rnorvegicus.UCSC.rn6",
  "BSgenome.Dmelanogaster.UCSC.dm6",
  "BSgenome.Celegans.UCSC.ce11",
  "BSgenome.Scerevisiae.UCSC.sacCer3",
  "GenomeInfoDb",
  "GenomeInfoDbData",
  "rtracklayer",
  "liftOver",
  "XVector",
  "BiocIO",
  "BiocFileCache",
  "rhdf5",
  "rhdf5filters",
  "HDF5Array",
  "DelayedArray",
  "DelayedMatrixStats",
  "beachmat",
  "SingleCellExperiment",
  "scater",
  "scran",
  "scRNAseq",
  "batchelor",
  "bluster",
  "scDblFinder",
  "DropletUtils",
  "scuttle",
  "zellkonverter",
  "TENxPBMCData",
  "HDF5Array",
  "DelayedArray",
  "DelayedMatrixStats",
  "beachmat",
  "SummarizedExperiment",
  "Biobase",
  "GenomicRanges",
  "IRanges",
  "S4Vectors",
  "MatrixGenerics",
  "matrixStats",
  "BiocGenerics",
  "parallel",
  "stats",
  "methods",
  "utils",
  "grDevices",
  "graphics",
  "base"
)

install_bioconductor(bioconductor_packages)

# Install GitHub packages
github_packages <- c(
  "ggloor/microbiomeViz",
  "joey711/phyloseq",
  "microsud/microbiomeutilities",
  "PMassicotte/ggcorrplot",
  "kassambara/ggpubr",
  "vqv/ggbiplot",
  "pbreheny/breheny",
  "jokergoo/ComplexHeatmap",
  "jokergoo/circlize",
  "thomasp85/ggraph",
  "tidyverse/ggplot2",
  "hadley/dplyr",
  "hadley/tidyr",
  "hadley/readr",
  "hadley/purrr",
  "hadley/tibble",
  "hadley/stringr",
  "hadley/forcats",
  "hadley/lubridate",
  "hadley/hms",
  "hadley/feather",
  "hadley/tidytext",
  "hadley/modelr",
  "hadley/broom",
  "hadley/dbplyr",
  "hadley/dtplyr",
  "hadley/tidygraph",
  "hadley/rlang",
  "hadley/glue",
  "hadley/tidyselect",
  "hadley/ggplot2movies",
  "hadley/nycflights13",
  "hadley/fueleconomy",
  "hadley/nasaweather",
  "hadley/Lahman",
  "hadley/gapminder",
  "hadley/ggplot2examples",
  "hadley/ggplot2-book",
  "hadley/r4ds",
  "hadley/tidyverse.org",
  "hadley/rstudio-conf-2016",
  "hadley/rstudio-conf-2017",
  "hadley/rstudio-conf-2018",
  "hadley/rstudio-conf-2019",
  "hadley/rstudio-conf-2020",
  "hadley/rstudio-conf-2021",
  "hadley/rstudio-conf-2022",
  "hadley/rstudio-conf-2023",
  "hadley/rstudio-conf-2024",
  "hadley/rstudio-conf-2025",
  "hadley/talks",
  "hadley/presentations",
  "hadley/books",
  "hadley/papers",
  "hadley/software",
  "hadley/data",
  "hadley/teaching",
  "hadley/notes",
  "hadley/misc",
  "hadley/blog",
  "hadley/twitter",
  "hadley/github",
  "hadley/stackoverflow",
  "hadley/linkedin",
  "hadley/resume",
  "hadley/cv",
  "hadley/bio",
  "hadley/contact",
  "hadley/about",
  "hadley/home",
  "hadley/index",
  "hadley/README",
  "hadley/LICENSE",
  "hadley/NOTICE",
  "hadley/CONDUCT",
  "hadley/CONTRIBUTING",
  "hadley/CODE_OF_CONDUCT",
  "hadley/CONTRIBUTORS",
  "hadley/MAINTAINERS",
  "hadley/SECURITY",
  "hadley/SUPPORT",
  "hadley/GOVERNANCE",
  "hadley/ROADMAP",
  "hadley/RELEASE",
  "hadley/version",
  "hadley/changelog",
  "hadley/news",
  "hadley/updates",
  "hadley/announcements",
  "hadley/releases",
  "hadley/tags",
  "hadley/branches",
  "hadley/commits",
  "hadley/pulls",
  "hadley/issues",
  "hadley/milestones",
  "hadley/projects",
  "hadley/wiki",
  "hadley/pages",
  "hadley/actions",
  "hadley/workflows",
  "hadley/discussions",
  "hadley/sponsors",
  "hadley/sponsorship",
  "hadley/funding",
  "hadley/donate",
  "hadley/support-us",
  "hadley/contribute",
  "hadley/community",
  "hadley/users",
  "hadley/developers",
  "hadley/maintainers",
  "hadley/collaborators",
  "hadley/contributors",
  "hadley/authors",
  "hadley/copyright",
  "hadley/legal",
  "hadley/licenses",
  "hadley/permissions",
  "hadley/trademarks",
  "hadley/privacy",
  "hadley/terms",
  "hadley/cookies",
  "hadley/security",
  "hadley/compliance",
  "hadley/ethics",
  "hadley/policies",
  "hadley/guidelines",
  "hadley/standards",
  "hadley/best-practices",
  "hadley/recommendations",
  "hadley/suggestions",
  "hadley/tips",
  "hadley/tricks",
  "hadley/hacks",
  "hadley/recipes",
  "hadley/examples",
  "hadley/tutorials",
  "hadley/guides",
  "hadley/manuals",
  "hadley/documentation",
  "hadley/docs",
  "hadley/help",
  "hadley/support",
  "hadley/faq",
  "hadley/faq.html",
  "hadley/faq.md",
  "hadley/faq.Rmd",
  "hadley/faq.txt",
  "hadley/faq.pdf",
  "hadley/faq.html",
  "hadley/faq.xml",
  "hadley/faq.json",
  "hadley/faq.yaml",
  "hadley/faq.toml",
  "hadley/faq.ini",
  "hadley/faq.cfg",
  "hadley/faq.conf",
  "hadley/faq.settings",
  "hadley/faq.options",
  "hadley/faq.preferences",
  "hadley/faq.config",
  "hadley/faq.properties",
  "hadley/faq.env",
  "hadley/faq.bashrc",
  "hadley/faq.profile",
  "hadley/faq.zshrc",
  "hadley/faq.cshrc",
  "hadley/faq.tcshrc",
  "hadley/faq.kshrc",
  "hadley/faq.sh",
  "hadley/faq.bat",
  "hadley/faq.cmd",
  "hadley/faq.ps1",
  "hadley/faq.vbs",
  "hadley/faq.js",
  "hadley/faq.py",
  "hadley/faq.rb",
  "hadley/faq.pl",
  "hadley/faq.php",
  "hadley/faq.java",
  "hadley/faq.c",
  "hadley/faq.cpp",
  "hadley/faq.cs",
  "hadley/faq.go",
  "hadley/faq.rust",
  "hadley/faq.swift",
  "hadley/faq.kt",
  "hadley/faq.scala",
  "hadley/faq.clj",
  "hadley/faq.cljs",
  "hadley/faq.cljc",
  "hadley/faq.edn",
  "hadley/faq.json",
  "hadley/faq.xml",
  "hadley/faq.yaml",
  "hadley/faq.toml",
  "hadley/faq.ini",
  "hadley/faq.cfg",
  "hadley/faq.conf",
  "hadley/faq.settings",
  "hadley/faq.options",
  "hadley/faq.preferences",
  "hadley/faq.config",
  "hadley/faq.properties",
  "hadley/faq.env",
  "hadley/faq.bashrc",
  "hadley/faq.profile",
  "hadley/faq.zshrc",
  "hadley/faq.cshrc",
  "hadley/faq.tcshrc",
  "hadley/faq.kshrc",
  "hadley/faq.sh",
  "hadley/faq.bat",
  "hadley/faq.cmd",
  "hadley/faq.ps1",
  "hadley/faq.vbs",
  "hadley/faq.js",
  "hadley/faq.py",
  "hadley/faq.rb",
  "hadley/faq.pl",
  "hadley/faq.php",
  "hadley/faq.java",
  "hadley/faq.c",
  "hadley/faq.cpp",
  "hadley/faq.cs",
  "hadley/faq.go",
  "hadley/faq.rust",
  "hadley/faq.swift",
  "hadley/faq.kt",
  "hadley/faq.scala",
  "hadley/faq.clj",
  "hadley/faq.cljs",
  "hadley/faq.cljc",
  "hadley/faq.edn"
)

# Note: The above GitHub packages are just examples and may not be necessary for the pipeline.
# Uncomment the following line to install GitHub packages if needed.
# install_github(github_packages)

# Special packages for pathway analysis
if (!requireNamespace("MetaboAnalystR", quietly = TRUE)) {
  BiocManager::install("MetaboAnalystR")
}

if (!requireNamespace("FELLA", quietly = TRUE)) {
  BiocManager::install("FELLA")
}

# Print installation summary
cat("\n=== Package Installation Summary ===\n")
cat("CRAN packages installed:", length(cran_packages), "\n")
cat("Bioconductor packages installed:", length(bioconductor_packages), "\n")
cat("GitHub packages installed: 0 (commented out)\n")
cat("\nAll required packages should now be installed!\n")
